include/unifex/finally.hpp

```diff
@@ -0,0 +1,679 @@
+        unifex::set_done(static_cast<Receiver&&>(op->receiver_));
+      }
+
+      template <
+          typename Error,
+          std::enable_if_t<
+              std::is_invocable_v<decltype(unifex::set_error), Receiver, Error>,
+              int> = 0>
+      void set_error(Error&& error) && noexcept {
+        auto* op = op_;
+        op->completionDoneOp_.destruct();
+        unifex::set_error(
+            static_cast<Receiver&&>(op->receiver_),
+            static_cast<Error&&>(error));
+      }
+
+      void set_done() && noexcept {
+          auto* op = op_;
+        op->completionDoneOp_.destruct();
+        unifex::set_done(static_cast<Receiver&&>(op->receiver_));
+      }
+
+      template <
+          typename CPO,
+          typename... Args,
+          std::enable_if_t<!is_receiver_cpo_v<CPO>, int> = 0>
+      friend auto tag_invoke(
+          CPO cpo,
+          const finally_done_receiver& r,
+          Args&&... args) noexcept(std::
+                                       is_nothrow_invocable_v<
+                                           CPO,
+                                           const Receiver&,
+                                           Args...>)
+          -> std::invoke_result_t<CPO, const Receiver&, Args...> {
+        return static_cast<CPO&&>(cpo)(
+            static_cast<const Receiver&>(r.op_->receiver_),
+            static_cast<Args&&>(args)...);
+      }
+
+      template <typename Func>
+      friend void tag_invoke(
+          tag_t<visit_continuations>,
+          const finally_done_receiver& r,
+          Func&& func) {
+        std::invoke(func, static_cast<const Receiver&>(r.op_->receiver_));
+      }
+
+      friend unstoppable_token
+      tag_invoke(tag_t<get_stop_token>, const finally_done_receiver&) noexcept {
+        return {};
+      }
+
+    private:
+      operation_type* op_;
+    };
+
+    template <
+        typename SourceSender,
+        typename CompletionSender,
+        typename Receiver>
+    class finally_receiver {
+      using operation_type =
+          finally_operation<SourceSender, CompletionSender, Receiver>;
+
+    public:
+      explicit finally_receiver(operation_type* op) noexcept : op_(op) {}
+
+      finally_receiver(finally_receiver&& other) noexcept
+        : op_(std::exchange(other.op_, nullptr)) {}
+
+      template <
+          typename... Values,
+          std::enable_if_t<
+              std::is_invocable_v<
+                  decltype(unifex::set_value),
+                  Receiver,
+                  std::decay_t<Values>...>,
+              int> = 0>
+      void set_value(Values&&... values) && noexcept {
+        auto* op = op_;
+        auto& valueStorage =
+            op->value_.template get<std::tuple<std::decay_t<Values>...>>();
+        try {
+          valueStorage.construct(static_cast<Values&&>(values)...);
+        } catch (...) {
+          std::move(*this).set_error(std::current_exception());
+          return;
+        }
```

Comment:
I don't love the contortions you have to go through to make your code exception safe when using `manual_lifetime`. Had you opted to use `std::optional` instead with the implicit-convertibility hack for non-movable types, this would be much simpler.